src/mito2/src/sst/index.rs

```diff
@@ -12,159 +12,155 @@
+use puffin_manager::{SstPuffinManager, SstPuffinWriter};
+use statistics::{ByteCount, RowCount};
 use store_api::metadata::RegionMetadataRef;
-use store_api::storage::RegionId;
+use store_api::storage::{ColumnId, RegionId};
 
+use crate::access_layer::OperationType;
+use crate::config::InvertedIndexConfig;
 use crate::metrics::INDEX_CREATE_MEMORY_USAGE;
 use crate::read::Batch;
 use crate::region::options::IndexOptions;
 use crate::sst::file::FileId;
 use crate::sst::index::intermediate::IntermediateManager;
+use crate::sst::index::inverted_index::creator::SstIndexCreator as InvertedIndexer;
+
+/// Output of the index creation.
+#[derive(Debug, Clone, Default)]
+pub struct IndexOutput {
+    /// Size of the file.
+    pub file_size: u64,
+    /// Inverted index output.
+    pub inverted_index: InvertedIndexOutput,
+}
 
-const INDEX_BLOB_TYPE: &str = "greptime-inverted-index-v1";
+/// Output of the inverted index creation.
+#[derive(Debug, Clone, Default)]
+pub struct InvertedIndexOutput {
+    /// Whether the index is available.
+    pub available: bool,
+    /// Size of the index.
+    pub index_size: ByteCount,
+    /// Number of rows in the index.
+    pub row_count: RowCount,
+    /// Available columns in the index.
+    pub columns: Vec<ColumnId>,
+}
 
 /// The index creator that hides the error handling details.
 #[derive(Default)]
 pub struct Indexer {
     file_id: FileId,
     region_id: RegionId,
-    inner: Option<SstIndexCreator>,
     last_memory_usage: usize,
+
+    inverted_indexer: Option<InvertedIndexer>,
+    puffin_writer: Option<SstPuffinWriter>,
 }
 
 impl Indexer {
-    /// Update the index with the given batch.
+    /// Updates the index with the given batch.
     pub async fn update(&mut self, batch: &Batch) {
-        if let Some(creator) = self.inner.as_mut() {
-            if let Err(err) = creator.update(batch).await {
-                if cfg!(any(test, feature = "test")) {
-                    panic!(
-                        "Failed to update index, region_id: {}, file_id: {}, err: {}",
-                        self.region_id, self.file_id, err
-                    );
-                } else {
-                    warn!(
-                        err; "Failed to update index, skip creating index, region_id: {}, file_id: {}",
-                        self.region_id, self.file_id,
-                    );
-                }
-
-                // Skip index creation if error occurs.
-                self.inner = None;
-            }
-        }
+        self.do_update(batch).await;
 
-        if let Some(creator) = self.inner.as_ref() {
-            let memory_usage = creator.memory_usage();
-            INDEX_CREATE_MEMORY_USAGE.add(memory_usage as i64 - self.last_memory_usage as i64);
-            self.last_memory_usage = memory_usage;
-        } else {
-            INDEX_CREATE_MEMORY_USAGE.sub(self.last_memory_usage as i64);
-            self.last_memory_usage = 0;
-        }
```

Comment:
So this is removed?

Reply:
If `creator` is `None`, The `0` memory usage acts the same for two branches.